R/preprocess.R

```diff
@@ -233,51 +345,173 @@ enw_reporting_triangle_to_long <- function(obs) {
+#' data frames as well as variables containing metadata. These are:
+#'  - `obs`: (observations with the addition of empirical reporting proportions
+#'  and and restricted to the specified maximum delay).
+#' - `new_confirm`: Incidence of notifications by reference and report date.
+#' Empirical reporting distributions are also added.
+#' - `latest`: The latest available observations.
+#' - `reporting_triangle`: Incident observations by report and reference date in
+#' the standard reporting triangle matrix format.
+#' - `metareference`: Metadata reference dates derived from observations.
+#' - `metrareport`: Metadata for report dates.
+#' - `time`: Numeric, number of timepoints in the data.
+#' - `snapshots`: Numeric, number of available data snapshots to use for
+#'  nowcasting.
+#' - `groups`: Numeric, Number of groups/strata in the supplied observations
+#'  (set using `by`).
+#' - `max_delay`: Numeric, the maximum delay in the processed data
+#' - `max_date`: The maximum available report date.
 #'
 #' @family preprocess
+#' @inheritParams enw_new_reports
 #' @export
 #' @importFrom data.table as.data.table data.table
+#' @examples
+#' library(data.table)
+#'
+#' # Filter example hospitalisation data to be natioanl and over all ages
+#' nat_germany_hosp <- germany_covid19_hosp[location == "DE"]
+#' nat_germany_hosp <- nat_germany_hosp[age_group %in% "00+"]
+#'
+#' # Preprocess with default settings
+#' pobs <- enw_preprocess_data(nat_germany_hosp)
+#' pobs
+#'
+#' # Preprocess using exclusion beyond the maximum delay and a max delay of 10
+#' pobs_exclude <- enw_preprocess_data(
+#'   nat_germany_hosp,
+#'   max_delay = 10, max_delay_strat = "exclude"
+#' )
+#' pobs_exclude
+#'
+#' # Preprocess all data
+#' pobs_all <- enw_preprocess_data(
+#'   germany_covid19_hosp,
+#'   by = c("location", "age_group")
+#' )
+#' pobs_all
 enw_preprocess_data <- function(obs, by = c(), max_delay = 20,
-                                rep_holidays = c(), ref_holidays = c(),
-                                min_report_date, set_negatives_to_zero = TRUE) {
+                                max_delay_strat = "add_to_max_delay",
+                                holidays = c(), set_negatives_to_zero = TRUE) {
+  max_delay_strat <- match.arg(
+    max_delay_strat,
+    choices = c("exclude", "add_to_max_delay")
+  )
   obs <- data.table::as.data.table(obs)
   obs <- obs[order(reference_date)]
 
-  if (!missing(min_report_date)) {
-    obs <- obs[report_date >= min_report_date]
-  }
-
-  # assign groups
   obs <- enw_assign_group(obs, by = by)
+  obs <- enw_add_max_reported(obs)
+  obs <- enw_add_delay(obs)
 
-  # filter by maximum report date
-  obs <- obs[, .SD[report_date <= (reference_date + max_delay - 1)],
-    by = c("reference_date", "group")
-  ]
+  if (max_delay_strat %in% "add_to_max_delay") {
+    obs[
+      report_date == (reference_date + max_delay - 1),
+      confirm := max_confirm,
+      by = "group"
+    ]
+  }
 
```

Comment:
As mentioned earlier, I am still confused by the current definition of `max_delay` here. As I understand, we are still excluding the maximum delay and only modeling until the day _before_ the maximum delay. Or am I getting something wrong?

Reply:
So I thought it should be the maximum delay assuming it is defined from 0 which is what we have in the stan code etc. That being said very happy to change that definition.